File: src/ruby/test/DataDictionaryTestCase.rb
require 'quickfix_ruby'
require 'test/unit'

class TooHigh < Quickfix::StringField
	def initialize(data = nil)
		if( data == nil )
			super(501)
		else
			super(501, data)
		end
	end
end

class DataDictionaryTestCase < Test::Unit::TestCase

	def setup
		@object = Quickfix::DataDictionary.new()
		assert_not_nil( @object )
	end

	def test_addMsgType
		assert( !@object.isMsgType("A") )
		@object.addMsgType("A")
		assert( @object.isMsgType("A") )
	end

	def test_addMsgField
		assert( !@object.isMsgField( "A", 10 ) )
		assert( !@object.isMsgField( "Z", 50 ) )
		@object.addMsgField( "A", 10 )
		@object.addMsgField( "Z", 50 )
		assert( @object.isMsgField( "A", 10 ) )
		assert( @object.isMsgField( "Z", 50 ) )
		assert( !@object.isMsgField( "A", 50 ) )
		assert( !@object.isMsgField( "Z", 10 ) )
	end

	def test_addHeaderField
		assert( !@object.isHeaderField( 56 ) )
		assert( !@object.isHeaderField( 49 ) )
		@object.addHeaderField( 56, true )
		@object.addHeaderField( 49, true )
		assert( @object.isHeaderField( 56 ) )
		assert( @object.isHeaderField( 49 ) )
	end

	def test_addTrailerField
		assert( !@object.isTrailerField( 10 ) )
		@object.addTrailerField( 10, true )
		assert( @object.isTrailerField( 10 ) )
	end

	def test_addRequiredField
		assert( !@object.isRequiredField( "A", 10 ) )
		assert( !@object.isRequiredField( "Z", 50 ) )
		@object.addRequiredField( "A", 10 )
		@object.addRequiredField( "Z", 50 )
		assert( @object.isRequiredField( "A", 10 ) )
		assert( @object.isRequiredField( "Z", 50 ) )
		assert( !@object.isRequiredField( "A", 50 ) )
		assert( !@object.isRequiredField( "Z", 10 ) )
	end

	def test_addFieldValue
		assert( !@object.isFieldValue( 12, "f" ) )
		assert( !@object.isFieldValue( 12, "g" ) )
		assert( !@object.isFieldValue( 15, "1" ) )
		assert( !@object.isFieldValue( 18, "2" ) )
		assert( !@object.isFieldValue( 167, "FUT" ) )

		@object.addFieldValue( 12, "f" )
		@object.addFieldValue( 12, "g" )
		@object.addFieldValue( 15, "1" )
		@object.addFieldValue( 18, "2" )
		@object.addFieldValue( 167, "FUT" )

		assert( @object.isFieldValue( 12, "f" ) )
		assert( @object.isFieldValue( 12, "g" ) )
		assert( @object.isFieldValue( 15, "1" ) )
		assert( @object.isFieldValue( 18, "2" ) )
		assert( @object.isFieldValue( 167, "FUT" ) )
	end

	def test_addGroup
		@object.setVersion( "FIX.4.2" )

		group1 = Quickfix::DataDictionary.new
		group1.addMsgType( "1" )
		group2 = Quickfix::DataDictionary.new
		group2.addMsgType( "2" )
		group3 = Quickfix::DataDictionary.new
		group3.addMsgType( "3" )

		@object.addGroup( "A", 100, 101, group1 )
		@object.addGroup( "A", 200, 201, group2 )
		@object.addGroup( "A", 300, 301, group3 )

		result = @object.getGroup( "A", 100 )
		assert_equal( 101, result[0] )
		assert( result[1].isMsgType( "1" ) )

		result = @object.getGroup( "A", 200 )
		assert_equal( 201, result[0] )
		assert( result[1].isMsgType( "2" ) )

		result = @object.getGroup( "A", 300 )
		assert_equal( 301, result[0] )
		assert( result[1].isMsgType( "3" ) )
	end

	def test_addFieldName
		@object.setVersion( "FIX.4.2" )

		@object.addFieldName( 1, "Account" )
		@object.addFieldName( 11, "ClOrdID" )
		@object.addFieldName( 8, "BeginString" )

		s = String.new
		@object.getFieldName(1)
                assert_equal( "Account", @object.getFieldName(1) )
		assert_equal( 1, @object.getFieldTag("Account") )
		assert_equal( "ClOrdID", @object.getFieldName(11) )
		assert_equal( 11, @object.getFieldTag("ClOrdID") )
		assert_equal( "BeginString", @object.getFieldName(8) )
		assert_equal( 8, @object.getFieldTag("BeginString") )
	end

	def test_addValueName
		@object.setVersion( "FIX.4.2" )

		@object.addValueName( 12, "0", "VALUE_12_0" )
		@object.addValueName( 12, "B", "VALUE_12_B" )
		@object.addValueName( 23, "BOO", "VALUE_23_BOO" )

		assert_equal( "VALUE_12_0", @object.getValueName(12, "0") )
		assert_equal( "VALUE_12_B", @object.getValueName(12, "B") )
		assert_equal( "VALUE_23_BOO", @object.getValueName(23, "BOO") )
	end

	def test_checkValidTagNumber
		@object.setVersion( Quickfix::BeginString_FIX40() )
		@object.addField( Quickfix::BeginString.field )
		@object.addField( Quickfix::BodyLength.field )
		@object.addField( Quickfix::MsgType.field )
		@object.addField( Quickfix::CheckSum.field )
		@object.addField( Quickfix::TestReqID.field )
		@object.addMsgType( Quickfix::MsgType_TestRequest() )
		@object.addMsgField( Quickfix::MsgType_TestRequest(), Quickfix::TestReqID.field )

		testReqID = Quickfix::TestReqID.new( "1" )
		message = Quickfix40::TestRequest.new()
		message.setField( testReqID )
		message.setField( TooHigh.new( "value" ) )

		begin
			@object.validate( message )
			assert( false )
		rescue Quickfix::Exception
		end

		@object.addField( 501 )
		@object.addMsgField( Quickfix::MsgType_TestRequest(), 501 )
		begin
			@object.validate( message )
		rescue Quickfix::Exception
			assert( false )
		end

		message.setField( 5000, "value" )
		begin
			@object.validate( message )
			assert( false )
		rescue Quickfix::Exception
		end

		@object.checkUserDefinedFields( false )
		begin
			@object.validate( message )
		rescue Quickfix::Exception
			assert( false )
		end
	end

	def test_checkHasValue
		message = Quickfix40::TestRequest.new
		message.setField( Quickfix::TestReqID.new )

		begin
			@object.validate( message )
			assert( false )
		rescue Quickfix::Exception
		end
	end

	def test_checkIsInMessage
		@object.setVersion( Quickfix::BeginString_FIX40() )
		@object.addField( Quickfix::BeginString.field )
		@object.addField( Quickfix::BodyLength.field )
		@object.addField( Quickfix::MsgType.field )
		@object.addField( Quickfix::CheckSum.field )
		@object.addField( Quickfix::TestReqID.field )
		@object.addField( Quickfix::Symbol.field )
		@object.addMsgType( Quickfix::MsgType_TestRequest() )
		@object.addMsgField( Quickfix::MsgType_TestRequest(), Quickfix::TestReqID.field )

		testReqID = Quickfix::TestReqID.new( "1" )

		message = Quickfix40::TestRequest.new
		message.setField( testReqID )
  		begin
			@object.validate( message )
		rescue Quickfix::Exception
			assert( false )
		end

		message.setField( Quickfix::Symbol.new("MSFT") )
		begin
			@object.validate( message )
			assert( false )
		rescue Quickfix::Exception
		end
	end

	def test_checkHasRequired
		@object.setVersion( Quickfix::BeginString_FIX40() )
		@object.addField( Quickfix::BeginString.field )
		@object.addField( Quickfix::BodyLength.field )
		@object.addField( Quickfix::MsgType.field )
		@object.addField( Quickfix::SenderCompID.field )
		@object.addField( Quickfix::TargetCompID.field )
		@object.addHeaderField( Quickfix::SenderCompID.field, true )
		@object.addHeaderField( Quickfix::TargetCompID.field, false )
		@object.addField( Quickfix::CheckSum.field )
		@object.addField( Quickfix::TestReqID.field )
		@object.addMsgType( Quickfix::MsgType_TestRequest() )
		@object.addMsgField( Quickfix::MsgType_TestRequest(), Quickfix::TestReqID.field )
		@object.addRequiredField( Quickfix::MsgType_TestRequest(), Quickfix::TestReqID.field )

		message = Quickfix40::TestRequest.new
		begin
			@object.validate( message )
			assert( false )
		rescue Quickfix::Exception
		end

		message.getHeader().setField( Quickfix::SenderCompID.new("SENDER") )
		begin
			@object.validate( message )
			assert( false )
		rescue Quickfix::Exception
		end

		message.setField( Quickfix::TestReqID.new( "1" ) )
		begin
			@object.validate( message )
		rescue Quickfix::Exception
			assert( false )
		end

		message.getHeader().removeField( Quickfix::SenderCompID.field )
		message.setField( Quickfix::SenderCompID.new("SENDER") )
		begin 
			@object.validate( message )
			assert( false )
		rescue Quickfix::Exception
		end
	end
end

File: src/python/test/DataDictionaryTestCase.py
import unittest

import quickfix as fix
import quickfix40 as fix40


class TooHigh(fix.StringField):

    def __init__(self, data=None):
        if data is None:
            fix.StringField.__init__(self, 501)
        else:
            fix.StringField.__init__(self, 501, data)


class DataDictionaryTestCase(unittest.TestCase):

    def setUp(self):
        self.object = fix.DataDictionary()

    def test_addMsgType(self):
        self.assertTrue(self.object.isMsgType("A") == 0)
        self.object.addMsgType("A")
        self.assertTrue(self.object.isMsgType("A"))

    def test_addMsgField(self):
        self.assertTrue(self.object.isMsgField("A", 10) == 0)
        self.assertTrue(self.object.isMsgField("Z", 50) == 0)
        self.object.addMsgField("A", 10)
        self.object.addMsgField("Z", 50)
        self.assertTrue(self.object.isMsgField("A", 10))
        self.assertTrue(self.object.isMsgField("Z", 50))
        self.assertTrue(self.object.isMsgField("A", 50) == 0)
        self.assertTrue(self.object.isMsgField("Z", 10) == 0)

    def test_addHeaderField(self):
        self.assertTrue(self.object.isHeaderField(56) == 0)
        self.assertTrue(self.object.isHeaderField(49) == 0)
        self.object.addHeaderField(56, True)
        self.object.addHeaderField(49, True)
        self.assertTrue(self.object.isHeaderField(56))
        self.assertTrue(self.object.isHeaderField(49))

    def test_addTrailerField(self):
        self.assertTrue(self.object.isTrailerField(10) == 0)
        self.object.addTrailerField(10, True)
        self.assertTrue(self.object.isTrailerField(10))

    def test_addRequiredField(self):
        self.assertTrue(self.object.isRequiredField("A", 10) == 0)
        self.assertTrue(self.object.isRequiredField("Z", 50) == 0)
        self.object.addRequiredField("A", 10)
        self.object.addRequiredField("Z", 50)
        self.assertTrue(self.object.isRequiredField("A", 10))
        self.assertTrue(self.object.isRequiredField("Z", 50))
        self.assertTrue(self.object.isRequiredField("A", 50) == 0)
        self.assertTrue(self.object.isRequiredField("Z", 10) == 0)

    def test_addFieldValue(self):
        self.assertTrue(self.object.isFieldValue(12, "f") == 0)
        self.assertTrue(self.object.isFieldValue(12, "g") == 0)
        self.assertTrue(self.object.isFieldValue(15, "1") == 0)
        self.assertTrue(self.object.isFieldValue(18, "2") == 0)
        self.assertTrue(self.object.isFieldValue(167, "FUT") == 0)

        self.object.addFieldValue(12, "f")
        self.object.addFieldValue(12, "g")
        self.object.addFieldValue(15, "1")
        self.object.addFieldValue(18, "2")
        self.object.addFieldValue(167, "FUT")

        self.assertTrue(self.object.isFieldValue(12, "f"))
        self.assertTrue(self.object.isFieldValue(12, "g"))
        self.assertTrue(self.object.isFieldValue(15, "1"))
        self.assertTrue(self.object.isFieldValue(18, "2"))
        self.assertTrue(self.object.isFieldValue(167, "FUT"))

    def test_addGroup(self):
        self.object.setVersion("FIX.4.2")

        group1 = fix.DataDictionary()
        group1.addMsgType("1")
        group2 = fix.DataDictionary()
        group2.addMsgType("2")
        group3 = fix.DataDictionary()
        group3.addMsgType("3")

        self.object.addGroup("A", 100, 101, group1)
        self.object.addGroup("A", 200, 201, group2)
        self.object.addGroup("A", 300, 301, group3)

        d = fix.DataDictionary()
        delim = 0

        result = self.object.getGroup("A", 100, delim, d)
        self.assertEqual(101, result[0])
        self.assertTrue(d.isMsgType("1"))

        result = self.object.getGroup("A", 200, delim, d)
        self.assertEqual(201, result[0])
        self.assertTrue(d.isMsgType("2"))

        result = self.object.getGroup("A", 300, delim, d)
        self.assertEqual(301, result[0])
        self.assertTrue(d.isMsgType("3"))

    def test_addFieldName(self):
        self.object.setVersion("FIX.4.2")

        self.object.addFieldName(1, "Account")
        self.object.addFieldName(11, "ClOrdID")
        self.object.addFieldName(8, "BeginString")

        self.assertEqual("Account", self.object.getFieldName(1, "")[0])
        self.assertEqual(1, self.object.getFieldTag("Account", 0)[0])
        self.assertEqual("ClOrdID", self.object.getFieldName(11, "")[0])
        self.assertEqual(11, self.object.getFieldTag("ClOrdID", 0)[0])
        self.assertEqual("BeginString", self.object.getFieldName(8, "")[0])
        self.assertEqual(8, self.object.getFieldTag("BeginString", 0)[0])

    def test_addValueName(self):
        self.object.setVersion("FIX.4.2")

        self.object.addValueName(12, "0", "VALUE_12_0")
        self.object.addValueName(12, "B", "VALUE_12_B")
        self.object.addValueName(23, "BOO", "VALUE_23_BOO")

        self.assertEqual(
            "VALUE_12_0", self.object.getValueName(12, "0", "")[0])
        self.assertEqual(
            "VALUE_12_B", self.object.getValueName(12, "B", "")[0])
        self.assertEqual(
            "VALUE_23_BOO", self.object.getValueName(23, "BOO", "")[0])

    def test_checkValidTagNumber(self):
        self.object.setVersion(fix.BeginString_FIX40)
        self.object.addField(fix.BeginString().getTag())
        self.object.addField(fix.BodyLength().getTag())
        self.object.addField(fix.MsgType().getTag())
        self.object.addField(fix.CheckSum().getTag())
        self.object.addField(fix.TestReqID().getTag())
        self.object.addMsgType(fix.MsgType_TestRequest)
        self.object.addMsgField(
            fix.MsgType_TestRequest, fix.TestReqID().getTag())

        testReqID = fix.TestReqID("1")
        message = fix40.TestRequest()
        message.setField(testReqID)
        message.setField(TooHigh("value"))

        self.assertRaises(fix.Exception, self.object.validate, message)

        self.object.addField(501)
        self.object.addMsgField(fix.MsgType_TestRequest, 501)
        try:
            self.object.validate(message)
            self.assertTrue(1)
        except fix.Exception:
            self.assertTrue(0)

        message.setField(5000, "value")
        self.assertRaises(fix.Exception, self.object.validate, message)

        self.object.checkUserDefinedFields(False)
        try:
            self.object.validate(message)
            self.assertTrue(1)
        except fix.Exception:
            self.assertTrue(0)

    def test_checkHasValue(self):
        message = fix40.TestRequest()
        message.setField(fix.TestReqID())
        self.assertRaises(fix.Exception, self.object.validate, message)

    def test_checkIsInMessage(self):
        self.object.setVersion(fix.BeginString_FIX40)
        self.object.addField(fix.BeginString().getTag())
        self.object.addField(fix.BodyLength().getTag())
        self.object.addField(fix.MsgType().getTag())
        self.object.addField(fix.CheckSum().getTag())
        self.object.addField(fix.TestReqID().getTag())
        self.object.addField(fix.Symbol().getTag())
        self.object.addMsgType(fix.MsgType_TestRequest)
        self.object.addMsgField(
            fix.MsgType_TestRequest, fix.TestReqID().getTag())

        testReqID = fix.TestReqID("1")

        message = fix40.TestRequest()
        message.setField(testReqID)
        try:
            self.object.validate(message)
            self.assertTrue(1)
        except fix.Exception:
            self.assertTrue(0)

        message.setField(fix.Symbol("MSFT"))
        self.assertRaises(fix.Exception, self.object.validate, message)

    def test_checkHasRequired(self):
        self.object.setVersion(fix.BeginString_FIX40)
        self.object.addField(fix.BeginString().getTag())
        self.object.addField(fix.BodyLength().getTag())
        self.object.addField(fix.MsgType().getTag())
        self.object.addField(fix.SenderCompID().getTag())
        self.object.addField(fix.TargetCompID().getTag())
        self.object.addHeaderField(fix.SenderCompID().getTag(), True)
        self.object.addHeaderField(fix.TargetCompID().getTag(), False)
        self.object.addField(fix.CheckSum().getTag())
        self.object.addField(fix.TestReqID().getTag())
        self.object.addMsgType(fix.MsgType_TestRequest)
        self.object.addMsgField(
            fix.MsgType_TestRequest, fix.TestReqID().getTag())
        self.object.addRequiredField(
            fix.MsgType_TestRequest, fix.TestReqID().getTag())

        message = fix40.TestRequest()
        self.assertRaises(fix.Exception, self.object.validate, message)

        message.getHeader().setField(fix.SenderCompID("SENDER"))
        self.assertRaises(fix.Exception, self.object.validate, message)

        message.setField(fix.TestReqID("1"))
        try:
            self.object.validate(message)
            self.assertTrue(1)
        except fix.Exception:
            self.assertTrue(0)

        message.getHeader().removeField(fix.SenderCompID().getTag())
        message.setField(fix.SenderCompID("SENDER"))
        self.assertRaises(fix.Exception, self.object.validate, message)

if __name__ == '__main__':
    unittest.main(verbosity=2)
